### ISSUE TITLE
GC does sweeps even when stopped

### ISSUE THREAD

**Issue by GitSparTV** [LuaJIT/LuaJIT#675]
Hi, I'm currently making memory profiler module and noticed GC is making sweeps even when it's stopped.

My code is stopping GC, doing full GC cycle and then indexing `int [10][10]` cdata.
```lua
for a = 0, 9 do
	for b = 0, 9 do
		accumulator = accumulator + cdata[a][b]
	end
end
```

Since my profiler is hooked directly to allocator I see all memory requests and I was surprised to receive `nsize = 0`.
It's triggered from `cdata__index` and doing partial sweep.
![image](https://user-images.githubusercontent.com/5685050/112230750-7399f100-8c46-11eb-944a-b4377603d481.png)

But should it do so?

`collectgarbage("stop")` sets the `gc.threshold` to `LJ_MAX_MEM`, which doesn't met conditions of `lj_gc_check` macro.
However calling `collectgarbage()` overwrites this value [here](https://github.com/LuaJIT/LuaJIT/blob/v2.1/src/lj_gc.c#L787
). Is this correct?

**Comment by GitSparTV** [comment]
Here is reproducible snippet
```lua
local switch = false
local function GC() if switch then error("GC WAS TRIGGERED EVEN WHEN STOPPED") end end

collectgarbage("stop")
collectgarbage()
collectgarbage()
print("START")

switch = true
for i = 1, 100000 do
	local obj = newproxy(true)
	local m = getmetatable(obj)
	m.__gc = GC
end
switch = false

print("DELETING")
collectgarbage()
collectgarbage()
```

**Comment by GitSparTV** [comment]
In Lua 5.2 it seems to be fixed with this:
https://www.lua.org/source/5.2/lgc.c.html#setpause

**Comment by MikePall** [comment]
I think you already gave the answer: Lua 5.1 behaves the same. And somebody may rely on that, so I can't change that easily.

**Comment by GitSparTV** [comment]
Alright, I personally don't belive someone knows GC might work when you stop it, if you told it to stop, you don't expect it to work.

**Comment by MikePall** [comment]
I've certainly seen the pattern of `collectgarbage("stop")` ... do some work ... `collectgarbage()` (full GC and implicit restart) in game code.

But, you're right: nobody should do it that way. In case I ever get around to the GC rewrite, I would probably switch to the Lua 5.2+ API behavior.

**Comment by GitSparTV** [comment]
About the reliabilty, Lua has few bugs in 5.1 version which were fixed only in 5.2, but LuaJIT fixed them though

**Comment by MikePall** [comment]
We're not talking about a bug, but an API deficiency that people rely on.

The perpetual API breaks of Lua are one of the main reasons for the developer mass exodus from Lua. Though I guess it's more about the imprudent, inconsiderate and gratuitous breakages rather than the reasonable changes, like in this case.
